cpp/log/cluster_state_controller-inl.h

```diff
@@ -0,0 +1,203 @@
+    LOG(WARNING) << status;
+  }
+}
+
+
+template <class Logged>
+void ClusterStateController<Logged>::OnClusterStateUpdated(
+    const std::vector<Update<ct::ClusterNodeState>>& updates) {
+  std::unique_lock<std::mutex> lock(mutex_);
+  for (const auto& update : updates) {
+    const std::string& node_id(update.handle_.Entry().node_id());
+    if (update.exists_) {
+      all_node_states_[node_id] = update.handle_.Entry();
+    } else {
+      CHECK_EQ(1, all_node_states_.erase(node_id));
+    }
+  }
+
+  if (CalculateServingSTH(lock) && election_->IsMaster()) {
+    update_required_ = true;
+    update_required_cv_.notify_all();
+  }
+}
+
+
+template <class Logged>
+bool ClusterStateController<Logged>::CalculateServingSTH(
+    const std::unique_lock<std::mutex>& lock) {
+  VLOG(1) << "Calculating new ServingSTH...";
+  CHECK(lock.owns_lock());
+
+  // First, create a mapping of tree size to number of nodes at that size, and
+  // a mapping of the newst STH for any given size:
+  std::map<uint64_t, ct::SignedTreeHead> sth_by_size;
+  std::map<uint64_t, int> num_nodes_by_sth_size;
+  for (const auto& node : all_node_states_) {
+    if (node.second.has_newest_sth()) {
+      const uint64_t tree_size(node.second.newest_sth().tree_size());
+      num_nodes_by_sth_size[tree_size]++;
+      if (node.second.newest_sth().timestamp() >
+          sth_by_size[tree_size].timestamp()) {
+        sth_by_size[tree_size] = node.second.newest_sth();
+      }
+    }
+  }
+
+  // Next calculate the newest STH we've seen which satisfies the following
+  // criteria:
+  //   - at least min_serving_nodes_ have an STH at least as large
+  //   - at least min_serving_fraction_ have an STH at least as large
+  //   - not smaller than the current serving STH
+  int num_nodes_seen(0);
+  const int current_tree_size(calculated_serving_sth_.get() != nullptr
+                                  ? calculated_serving_sth_->tree_size()
+                                  : 0);
+
+  // Work backwards (from largest STH size) until we see that there's enough
+  // coverage (according to the criteria above) to serve an STH (or determine
+  // that there are insufficient nodes to serve anything.)
+  for (auto it = num_nodes_by_sth_size.rbegin();
+       it != num_nodes_by_sth_size.rend() && it->first >= current_tree_size;
+       ++it) {
+    // num_nodes_seen keeps track of the number of nodes we've seen so far (and
+    // since we're working from larger to smaller size STH, they should all be
+    // able to serve this [and smaller] STHs.)
+    num_nodes_seen += it->second;
+    const double serving_fraction(static_cast<double>(num_nodes_seen) /
+                                  all_node_states_.size());
+    if (serving_fraction >= min_serving_fraction_ &&
+        num_nodes_seen >= min_serving_nodes_) {
+      LOG(INFO) << "Can serve @" << it->first << " with " << num_nodes_seen
+                << " nodes (" << (serving_fraction * 100) << "% of cluster)";
```

Comment:
I used to prefer percentages, but I've been at Google long enough that now I prefer 0..1 intervals. :smiley: